### PR TITLE
SDK Upgrade for NDK - Native Crash c++

### PR DESCRIPTION
**Pull request by thinkocapo**
### Overall Goal
:confetti_ball: To show a Native Crash from a c++ plugin :confetti_ball:

> The NDK integration is packed with the SDK and enabled by default with each start of the SDK.

Sentry Android Documentation - https://docs.sentry.io/platforms/android/

### Goals
- [x]  update SDK
    - [x]  rip out old SDK
    - [x]  put in new Sdk - reference the slides
    - [x]  test if java errors are being surfaced and ANR
- [x]  add native crash button - [https://github.com/getsentry/sentry-android/search?
q=NativeSample&unscoped_q=NativeSample](https://github.com/getsentry/sentry-android/search?q=NativeSample&unscoped_q=NativeSample) from sentry-sample
^ not sure what this is doing, no cpp crash. it's just a java error?
- [x]  add GIF showcasing surfacing of native/NDK error

### Versions
This was last tested successfully on 01/02/20 with:
```
Android Studio 3.5.3
Build #AI-191.8026.42.35.6010548, built on November 15, 2019
JRE: 1.8.0_202-release-1483-b49-5587405 x86_64
JVM: OpenJDK 64-Bit Server VM by JetBrains s.r.o
macOS 10.14.4
```
**AVD** Nexus 5x API 29 x86
**Gradle** .gradle/.5.6.4

## Some Design Decisions
changed ANR time to 2seconds for faster demo
<meta-data android:name="io.sentry.anr.timeout-interval-mills" android:value="2000" />

not auto-initializing Sentry from AndroidManifest.xml anymore, so we can show `Sentry.init` in the demo and have access to its config, in MainActivity.java:
<meta-data android:name="io.sentry.auto-init" android:value="false" />

#### Design
Keep the titles between the README vs code vs Button titles consistent!?
e.g.
`ERROR 1 - DIVIDE BY 0` could be `DIVIDE BY 0 - Arithmetic Exception`? Because the error on Sentry.io says 'Arithmetic Exception'.

## Most Recent Test
01/02/20
Button1 - Arithmetic Exception https://sentry.io/organizations/testorg-az/projects/android/events/860b0c326d9649cc946a5b16f60f4ac8/

Button2 - NegativeArraySizeException https://sentry.io/organizations/testorg-az/projects/android/events/6bf0cb865c8e4b55aee1d3a0ab631b73/

Button3 - NumberFormatException https://sentry.io/organizations/testorg-az/projects/android/events/8ac22ab321fd4ff8af9124a36c8383b8/

Button4 - ANR https://sentry.io/organizations/testorg-az/projects/android/events/d6dae2eab9904f0b97092109a32a18b5/

Button5 - Native Exception https://sentry.io/organizations/testorg-az/projects/android/events/1587daa817f34f99f47e6e8a21a0112b/

## Next Pull Requests (Q1 2020)
- LinearLayout instead of ConstraintLayout in activity_main.xml?
- Animation for button onPush so you tell it was clicked
- ProGuard?
- Timeline of additional features coming out that we should prepare for sentry-demos?
- Professional Design from Dora and team on its way, implement it (for iOS too)

### CODE REVIEW

File: app/src/main/java/com/example/vu/android/MainActivity.java [hunk 1]
```diff
@@ -1,19 +1,22 @@
 package com.example.vu.android;
 
 import android.os.Bundle;
-import io.sentry.android.AndroidSentryClientFactory;
-import android.support.v7.app.AppCompatActivity;
+// are no longer relevant with io.sentry:sentry-android:2.0.+ ?
```
Comment (marandaneto) [inline app/src/main/java/com/example/vu/android/MainActivity.java]: no

File: app/src/main/cpp/native-sample.cpp [hunk 2]
```diff
@@ -0,0 +1,15 @@
+
+#include <jni.h>
+#include <android/log.h>
+
+#define TAG "sentry-android-sample"
+
+extern "C" {
+
+JNIEXPORT void JNICALL Java_io_sentry_sample_NativeSample_crash(JNIEnv *env, jclass cls) {
```
Comment (marandaneto) [inline app/src/main/cpp/native-sample.cpp]: package here should be related to your own package, unless your app is also called io.sentry.sample

Comment (thinkocapo) [reply app/src/main/cpp/native-sample.cpp]: we fixed it, thanks.

File: proguard-rules.pro [hunk 3]
```diff
@@ -0,0 +1,21 @@
+# Add project specific ProGuard rules here.
```
Comment (thinkocapo) [inline proguard-rules.pro]: @marandaneto can I leave this file here? I don't remember generating it. I don't think I'm using proguard right now, but probably will eventually. Can do it on a next branch or PullRequest, would it help add better context to the Event? https://docs.sentry.io/platforms/android/#proguard

Comment (marandaneto) [reply proguard-rules.pro]: yes, you can.
it's generated automatically and its empty, so no issues.

File: app/src/main/AndroidManifest.xml [hunk 4]
```diff
@@ -1,10 +1,14 @@
 <?xml version="1.1" encoding="utf-8"?>
 <manifest xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:tools="http://schemas.android.com/tools"
     package="com.example.vu.android">
 
-    <uses-permission android:name="android.permission.INTERNET" />
+<!--    <uses-permission android:name="android.permission.INTERNET" />-->
```
Comment (marandaneto) [inline app/src/main/AndroidManifest.xml]: you can remove this comment

File: app/src/main/java/com/example/vu/android/MainActivity.java [hunk 5]
```diff
@@ -1,108 +1,94 @@
...
+        SentryAndroid.init(this, options -> {
+            // This callback is used before the event is sent to Sentry.
+            // You can modify the event or, when returning null, also discard the event.
+            options.setBeforeSend((event, hint) -> {
+                if (SentryLevel.DEBUG.equals(event.getLevel()))
+                    return null;
+                else
+                    return event;
+            });
         });
```
Comment (marandaneto) [inline app/src/main/java/com/example/vu/android/MainActivity.java]: this should be called on `Application` class like https://github.com/getsentry/sentry-android/blob/1a8bbd7f4c529971179a69e65d07f41394797c06/sentry-sample/src/main/java/io/sentry/sample/MyApplication.java#L21

it works, but it's not the right place to do so

Comment (marandaneto) [reply app/src/main/java/com/example/vu/android/MainActivity.java]: after creating a custom `Application` class, you need to be set also in the `manifest`, see https://github.com/getsentry/sentry-android/blob/1a8bbd7f4c529971179a69e65d07f41394797c06/sentry-sample/src/main/AndroidManifest.xml#L11

Comment (thinkocapo) [reply app/src/main/java/com/example/vu/android/MainActivity.java]: Ok I made these changes and moved Sentry initialization to MyApplication.class (the Application class) and re-tested, all buttons still work.

File: app/src/main/java/com/example/vu/android/MyApplication.java [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+        //    https://developer.android.com/reference/android/os/StrictMode
+        //    StrictMode is a developer tool which detects things you might be doing by accident and
+        //    brings them to your attention so you can fix them.
+        if (BuildConfig.DEBUG) {
+            StrictMode.setThreadPolicy(
+                    new StrictMode.ThreadPolicy.Builder().detectAll().penaltyLog().build());
+
+            StrictMode.setVmPolicy(new StrictMode.VmPolicy.Builder().detectAll().penaltyLog().build());
+        }
+    }
```
Comment (marandaneto) [inline app/src/main/java/com/example/vu/android/MyApplication.java]: this is not really important for a demo, you can remove it

File: app/src/main/java/com/example/vu/android/MyApplication.java [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+import android.app.Application;
+import android.os.StrictMode;
+
+import io.sentry.android.core.SentryAndroid;
+import io.sentry.core.SentryLevel;
+
+public class MyApplication extends Application {
+    @Override
+    public void onCreate() {
+        districtMode();
```
Comment (marandaneto) [inline app/src/main/java/com/example/vu/android/MyApplication.java]: as the moved is gonna be removed, remove this line too

File: README.md [hunk 8]
```diff
@@ -21,27 +44,44 @@ See https://docs.sentry.io/clients/java/modules/android/ for more information.
     Tools -> Android -> Sync Project with Gradle Files
     ```
 
-4. Update the DSN in the `app/src/main/resources/sentry.properties` file
+4. Put your DSN key in `app/src/main/resources/sentry.properties`
```
Comment (marandaneto) [inline README.md]: DSN is supposed to be set here: `AndroidManifest.xml`, this path is for the old SDK.
We still use `sentry.properties` but only for the Gradle plugin, which is not being used on this demo.

File: app/src/main/java/com/example/vu/android/MyApplication.java (hunk 6 above)
Comment (thinkocapo) [reply app/src/main/java/com/example/vu/android/MyApplication.java]: done

File: README.md (hunk 8 above)
Comment (thinkocapo) [reply README.md]: done

File: app/src/main/java/com/example/vu/android/MyApplication.java (hunk 7 above)
Comment (thinkocapo) [reply app/src/main/java/com/example/vu/android/MyApplication.java]: done

**Pull request merged**